### PR TITLE
Enable Transformer LT search space for Dynamic Neural Architecture Search Toolkit

### PR DESCRIPTION
**Pull request by macsz**
## Type of Change

Feature. Extends current Dynamic Neural Architecture Search Toolkit set of supported search spaces with Transformer LT super-network for En-De language translation.

## Description

DyNAS-T (Dynamic Neural Architecture Search Toolkit) is a SuperNet NAS optimization package (part of Intel Neural Compressor) designed for finding the optimal Pareto front during neural architecture search while minimizing the number of search validation measurements. It supports single-/multi-/many-objective problems for a variety of domains supported. The system currently heavily utilizes the [pymoo](https://pymoo.org/) optimization library. Some of the key DyNAS-T features are:

* Automatic handling of super-network parameters for search and predictor training
* Genetic Algorithm (e.g., NSGA-II) multi-objective subnetworks
* LINAS (Lightweight Iterative Neural Architecture Search) accelerated search using approximate predictors
* Warm-start (transfer) search
* Search population statistical analysis

This PR extends supported search spaces with Transformer-based Language Translation (`transformer_lt_wmt_en_de`) for English and German languages. Implementation of the supernet is based on [Hardware Aware Transformers (HAT)](https://github.com/mit-han-lab/hardware-aware-transformers) by MIT HAN Lab.

## How has this PR been tested?

To run an example, trained supernet weights and preprocessed WMT En-De dataset is needed. Both can be downloaded from [Hardware Aware Transformers (HAT)](https://github.com/mit-han-lab/hardware-aware-transformers) repository.

* Script to download preprocessed dataset: [link](https://github.com/mit-han-lab/hardware-aware-transformers/blob/master/configs/wmt14.en-de/get_preprocessed.sh)
* Download trained supernet weights: [link](https://www.dropbox.com/s/pkdddxvvpw9a4vq/HAT_wmt14ende_super_space0.pt?dl=0)

Example code to test new functionality:

```python
config = NASConfig(approach='dynas', search_algorithm='nsga2')
config.dynas.supernet = 'transformer_lt_wmt_en_de'
config.seed = 42
config.dynas.metrics = ['acc', 'macs']

config.dynas.population = 50
config.dynas.num_evals = 500
config.dynas.batch_size = 64
config.dynas.results_csv_path = 'results.csv'
config.dynas.dataset_path = '/datasets/hat_dataset/data/binary/wmt16_en_de'
config.dynas.supernet_ckpt_path  ='/datasets/hat_dataset/HAT_wmt14ende_super_space0.pt'
agent = NAS(config)
results = agent.search()
```

## Dependency Change?

* `fairseq`
* `sacremoses`
* `torchprofile`

Signed-off-by: Maciej Szankin <maciej.szankin@intel.com>

### CODE REVIEW

File: examples/notebook/dynas/Supernet_NAS.ipynb [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "\n",
```
Comment (XinyuYe-Intel) [inline examples/notebook/dynas/Supernet_NAS.ipynb:13]: May be better to use separate notebook for different models?

**Comment by ftian1** [comment]
@macsz 

A major difference between INC 2.0 and 1.x is the user facing API design philosophy. INC 2.0 is Config class driven but INC 1.x is not. It causes many parameters in INC 1.x are set by object's attributes, but INC2.0 can set those info by the input parameters of the initialization func of Conf class.

so suggest to modify the below code from

```
config = NASConfig(approach='dynas', search_algorithm='nsga2')
config.dynas.supernet = 'transformer_lt_wmt_en_de'
config.seed = 42
config.dynas.metrics = ['acc', 'macs']

config.dynas.population = 50
config.dynas.num_evals = 500
config.dynas.batch_size = 64
config.dynas.results_csv_path = 'results.csv'
config.dynas.dataset_path = '/datasets/hat_dataset/data/binary/wmt16_en_de'
config.dynas.supernet_ckpt_path  ='/datasets/hat_dataset/HAT_wmt14ende_super_space0.pt'
agent = NAS(config)
results = agent.search()
```

to 

```
dynas_setting = Dynas(supernet='transformer_lt_wmt_en_de', metrics=..., population=..., ...)
config = NASConfig(approach='dynas', search_algorithm='nsga2', seed=42, setting=dynas_setting)
agent = NAS(config)
results = agent.search()
```

PS: you can decide it to be in 2.0 or future 2.1 release from resource view

**Comment by macsz** [comment]
> 

@ftian1 Just to be clear - the 2.0 v 2.1 you meant it for the API design philosophy or the Transformer LT search space in general?

**Comment by ftian1** [comment]
> @ftian1 Just to be clear - the 2.0 v 2.1 you meant it for the API design philosophy or the Transformer LT search space in general?

what I meant is INC 2.0 user facing APIs have been all revised to be config-driven. so you will see two sets of APIs: the new one under top dir/namespace, the old one under experimental dir/namespace. For this Dynas feature (not specific to this PR), currently it's under experiemntal dir. you can consider to have unified API like what we have in INC 2.0 when you think it's mature enough and ok to move to top namespace. 

so it's not a urgent request, you can decide when to update based on your resource.

File: examples/notebook/dynas/Supernet_NAS.ipynb (hunk 1 above)
Comment (macsz) [reply examples/notebook/dynas/Supernet_NAS.ipynb]: Done!

**Comment by macsz** [comment]
> @macsz
> 
> A major difference between INC 2.0 and 1.x is the user facing API design philosophy. INC 2.0 is Config class driven but INC 1.x is not. It causes many parameters in INC 1.x are set by object's attributes, but INC2.0 can set those info by the input parameters of the initialization func of Conf class.
> 
> so suggest to modify the below code from
> 
> ```
> config = NASConfig(approach='dynas', search_algorithm='nsga2')
> config.dynas.supernet = 'transformer_lt_wmt_en_de'
> config.seed = 42
> config.dynas.metrics = ['acc', 'macs']
> 
> config.dynas.population = 50
> config.dynas.num_evals = 500
> config.dynas.batch_size = 64
> config.dynas.results_csv_path = 'results.csv'
> config.dynas.dataset_path = '/datasets/hat_dataset/data/binary/wmt16_en_de'
> config.dynas.supernet_ckpt_path  ='/datasets/hat_dataset/HAT_wmt14ende_super_space0.pt'
> agent = NAS(config)
> results = agent.search()
> ```
> 
> to
> 
> ```
> dynas_setting = Dynas(supernet='transformer_lt_wmt_en_de', metrics=..., population=..., ...)
> config = NASConfig(approach='dynas', search_algorithm='nsga2', seed=42, setting=dynas_setting)
> agent = NAS(config)
> results = agent.search()
> ```
> 
> PS: you can decide it to be in 2.0 or future 2.1 release from resource view

@ftian1 `NASConfig` class does not have `setting` param. Has the API change you are referring to been merged, or is it available on some other branch? I merged master into this branch but getting error:

```
config = NASConfig(approach='dynas', search_algorithm='nsga2', setting=dynas_setting)
TypeError: __init__() got an unexpected keyword argument 'setting
```

Old config interface also stopped working after merging current changes from master.

**Comment by macsz** [comment]
> Old config interface also stopped working after merging current changes from master.

Current branch works after merging changes from master - what I reported above was a mistake in configuration on my part.

I've created a separate WIP PR for API change, as I am not sure what direction to take. @ftian1 could you take a look and help me with https://github.com/intel/neural-compressor/pull/246 if you have time? I understand the change can happen for 2.0 or 2.1, will try to make it for the 2.0, but if not we can work on it for 2.1. Should I work on the implementation around the `setting` parameter of the `NASConfig` class, or is it something that already has been done somewhere? Don't see it on master branch. Please advise :)

**Comment by chensuyue** [comment]
@macsz please check the CI failures.

**Comment by ftian1** [comment]
thanks Maciej. I will give comments on PR 246.

**Review by ftian1 — APPROVED** [review]

**Comment by chensuyue** [comment]
Please check CI issue.

**Comment by macsz** [comment]
@chensuyue are there any requirements.txt used just for running tests, or should I update the main requirements.txt? It fails on missing dependency (fairseq).

**Comment by macsz** [comment]
> @chensuyue are there any requirements.txt used just for running tests, or should I update the main requirements.txt? It fails on missing dependency (fairseq).

Nevermind, found it :)

**Comment by macsz** [comment]
Seems like remaining errors are unrelated to this PR. Mainy horovod and failed to download model from Tensorflow. Will merge master and re-run.

**Comment by macsz** [comment]
Current problem is with torch version. I do not have this issue when running locally, but also my testing environment is not complete (not full NC testing suite). What I am seeing right now is that if I leave test requirements as is, it will complain about lack of fairseq library. Fair. But when I add it to test requirements all IPEX-related tests fail due to pytorch version mismatch - torch 1.12 is required for IPEX, with fairseq we magically get upgraded to 1.13 on CI. Locally when I already have torch 1.12 and install fairseq torch never gets upgraded, not sure what could be the root cause here.

**Comment by macsz** [comment]
Failing on timeout now. Locally the same test passes.

**Review by XinyuYe-Intel — APPROVED** [review]

**Comment by ftian1** [comment]
@macsz we have root caused the timeout issue and fixed it. we also made rebase operation. now we are waiting for test report. if it passes, we will merge it today.

**Comment by macsz** [comment]
> @macsz we have root caused the timeout issue and fixed it. we also made rebase operation. now we are waiting for test report. if it passes, we will merge it today.

Thanks! Appreciate the help to rebase to speed things up! I will be monitoring the PR as well.

**Comment by chensuyue** [comment]
image
Failed with UT coverage regression. I think we can merge at first for code freeze.

**Comment by macsz** [comment]
Merged master & Removed cleanup in UTs that I added before you fixed the model cache problem. This triggered the CI to re-run the tests.

So we are ok with this PR once it passes? To avoid additional delays I won't be touching anything till you say so.

By the way - Where in the CI can I see visual test coverage, like the one you posted @chensuyue ? Tried decoding raw output but it was a little bit too much ...

**Comment by chensuyue** [comment]
1. Go to the artifacts.
image
2. Download the coverage report package.
image
3. Open the coverage compare html. 
image

**Comment by macsz** [comment]
Should we merge now for code freeze, as @chensuyue suggested? I will work on adding unit tests as a follow up to this PR.

**Pull request merged**